### PR TITLE
feature: allow tokens without sub and aud claims to be validated

### PR DESCRIPTION
**Pull request by ybelMekk**

### CODE REVIEW

File: token-validation-core/src/main/java/no/nav/security/token/support/core/validation/CasualJwtTokenValidator.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import com.nimbusds.jwt.proc.DefaultJWTClaimsVerifier;
+import com.nimbusds.jwt.proc.DefaultJWTProcessor;
+import com.nimbusds.jwt.proc.JWTClaimsSetVerifier;
+import no.nav.security.token.support.core.exceptions.JwtTokenValidatorException;
+
+import java.net.URL;
+import java.util.HashSet;
+import java.util.List;
+
+public class CasualJwtTokenValidator implements JwtTokenValidator {
```
Comment (tommytroen) [inline token-validation-core/src/main/java/no/nav/security/token/support/core/validation/CasualJwtTokenValidator.java:21]: Not sold on the name "casual" (naming is hard). I was thinking we might go for a "configurable" validator with sane defaults (as we briefly discussed in the mob prog session) - as such the class could be named `ConfigurableJwtTokenValidator` or something like that - and take a list of required claims from config (other than iss, nbf and exp).

File: token-validation-core/src/main/java/no/nav/security/token/support/core/validation/CasualJwtTokenValidator.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    private void verify(String tokenString, JWTClaimsSetVerifier<SecurityContext> jwtClaimsSetVerifier, JWSVerificationKeySelector<SecurityContext> keySelector) {
+        try {
+            ConfigurableJWTProcessor<SecurityContext> jwtProcessor = new DefaultJWTProcessor<>();
+            jwtProcessor.setJWSKeySelector(keySelector);
+            jwtProcessor.setJWTClaimsSetVerifier(jwtClaimsSetVerifier);
+            JWT token = parse(tokenString);
+            jwtProcessor.process(token, null);
+        } catch (Throwable t) {
+            throw new JwtTokenValidatorException("Token validation failed: " + t.getMessage());
```
Comment (tommytroen) [inline token-validation-core/src/main/java/no/nav/security/token/support/core/validation/CasualJwtTokenValidator.java:63]: Need to provide the original exception here, or the original cause is lost and all we got is the message

File: token-validation-core/src/main/java/no/nav/security/token/support/core/validation/CasualJwtTokenValidator.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        }
+    }
+
+    private JWT parse(String tokenString) {
+        try {
+            return JWTParser.parse(tokenString);
+        } catch (Throwable t) {
+            throw new JwtTokenValidatorException("Token verification failed: " + t.getMessage());
```
Comment (tommytroen) [inline token-validation-core/src/main/java/no/nav/security/token/support/core/validation/CasualJwtTokenValidator.java:71]: Same as above

File: token-validation-core/src/main/java/no/nav/security/token/support/core/validation/CasualJwtTokenValidator.java (hunk 1 above)
Comment (ybelMekk) [reply token-validation-core/src/main/java/no/nav/security/token/support/core/validation/CasualJwtTokenValidator.java:21]: Agree, as always, naming is hard. But seems like the better choice.

File: README.md [hunk 4]
```diff
@@ -238,6 +238,7 @@ Add the modules that you need as Maven dependencies.
 - **`no.nav.security.jwt.issuer.[issuer shortname].accepted_audience`** - The value of the audience (aud) claim in the JWT token. For OIDC it is the client ID of the client responsible for acquiring the token, in OAuth 2.0 it should be the identifier for you api.
 - **`no.nav.security.jwt.issuer.[issuer shortname].cookiename`** - The value of the cookie containing a ID token (not required, only necessary if your api receives calls directly from a browser)
 - **`no.nav.security.jwt.issuer.[issuer shortname].configurable_claim_validator`** - A more relaxed Jwt Validator, as the name indicates, more configurable, sub and aud is not validated by default.
+- **`no.nav.security.jwt.issuer.[issuer shortname].requred_claims`** - If using a more configurable Jwt Validator, other claims required can be added.
```
Comment (ybelMekk) [inline README.md:241]: Not sure about the description is right here.

File: token-validation-core/src/main/java/no/nav/security/token/support/core/validation/ConfigurableJwtTokenValidator.java [hunk 5]
```diff
@@ -59,16 +66,25 @@ private void verify(String tokenString, JWTClaimsSetVerifier<SecurityContext> jw
...
             return JWTParser.parse(tokenString);
-        } catch (Throwable t) {
-            throw new JwtTokenValidatorException("Token verification failed: " + t.getMessage());
+        } catch (ParseException e) {
+            throw new JwtTokenValidatorException("Token verification failed: " + e.getMessage(), e);
         }
     }
+
+    private List<String> add(List<String> providedClaims) {
+        if (providedClaims == null || providedClaims.isEmpty()) {
```
Comment (ybelMekk) [inline token-validation-core/src/main/java/no/nav/security/token/support/core/validation/ConfigurableJwtTokenValidator.java:83]: Is this check needed? or can the list be empty?

File: README.md [hunk 6]
```diff
@@ -236,7 +236,9 @@ Add the modules that you need as Maven dependencies.
...
 - **`no.nav.security.jwt.issuer.[issuer shortname].discoveryurl`** - The identity provider configuration/discovery endpoint (metadata)
 - **`no.nav.security.jwt.issuer.[issuer shortname].accepted_audience`** - The value of the audience (aud) claim in the JWT token. For OIDC it is the client ID of the client responsible for acquiring the token, in OAuth 2.0 it should be the identifier for you api.
-- **`no.nav.security.jwt.issuer.[issuer shortname].cookiename`** - The value of the cookie containing a ID token (not required, only neccessary if your api receives calls directly from a browser)
+- **`no.nav.security.jwt.issuer.[issuer shortname].cookiename`** - The value of the cookie containing a ID token (not required, only necessary if your api receives calls directly from a browser)
+- **`no.nav.security.jwt.issuer.[issuer shortname].configurable_claim_validator`** - A more relaxed Jwt Validator, as the name indicates, more configurable, sub and aud is not validated by default.
+- **`no.nav.security.jwt.issuer.[issuer shortname].required_claims`** - If using a more configurable Jwt Validator, other claims required can be added.
```
Comment (ybelMekk) [inline README.md:241]: probably some better description?

**Pull request merged**